File: scripts/generate_ai_descriptions.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { GoogleGenAI } from '@google/genai';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function processBatch(ai, queriesBatch, retries = 3) {
    const prompt = `
Tu es un expert métier et technique sur l'ERP Sage X3.
Voici une liste de requêtes SQL. Pour chacune, génère une description très courte (1 à 2 phrases) en français qui explique simplement à quoi sert cette extraction de données pour un utilisateur métier.

Voici les requêtes :
${queriesBatch.map(q => `ID: ${q.id} | Titre: ${q.title} | Domaine: ${q.domain} | Tables: ${(q.fields||[]).join(', ')}\nSQL: ${q.sql}`).join('\n\n')}

Tu dois OBLIGATOIREMENT répondre avec un objet JSON dont les clés sont les ID des requêtes et les valeurs sont les descriptions générées.
Exemple attendue : {"id1": "Description 1", "id2": "Description 2"}
`;

    try {
        const response = await ai.models.generateContent({
            model: 'gemini-3.6-flash', // On repasse sur le modèle officiel
            contents: prompt,
            config: {
                responseMimeType: "application/json"
            }
        });
        
        const text = response.text.trim();
        return JSON.parse(text);
    } catch (error) {
        const isTransient = error.status === 429 || error.status === 503 || (error.error && error.error.code === 503);
        if (isTransient && retries > 0) {
            const waitMs = (error.status === 429) ? 60000 : 15000;
            console.log(` [Serveur saturé / Quota (${error.status || '503'}), nouvelle tentative dans ${waitMs/1000}s...]`);
            await delay(waitMs);
            return processBatch(ai, queriesBatch, retries - 1);
        }
        console.error(`\nErreur lors du traitement du lot: ${error.message}`);
        return null;
    }
}

async function main() {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) {
        console.error("Erreur : La variable d'environnement GEMINI_API_KEY n'est pas définie.");
        process.exit(1);
    }

    const ai = new GoogleGenAI({ apiKey: apiKey });
    
    const catalogPath = path.join(__dirname, '..', 'src', 'lib', 'data', 'queriesCatalog.json');
    if (!fs.existsSync(catalogPath)) {
        console.error(`Erreur: Le fichier ${catalogPath} est introuvable.`);
        process.exit(1);
    }

    console.log("Chargement du catalogue des requêtes...");
    const rawData = fs.readFileSync(catalogPath, 'utf-8');
    const catalog = JSON.parse(rawData);

    // On isole uniquement les requêtes qui n'ont pas encore de description personnalisée
    const pendingQueries = catalog.filter(q => q.description && q.description.startsWith("Extraction de données basée sur les tables"));
    
    console.log(`${pendingQueries.length} requêtes restantes à traiter.`);
    console.log("Pour contourner la limite de 20 requêtes/jour de Google, on va les traiter par lots de 15 !\n");

    const BATCH_SIZE = 15;
    let modificationsCount = 0;

    for (let i = 0; i < pendingQueries.length; i += BATCH_SIZE) {
        const batch = pendingQueries.slice(i, i + BATCH_SIZE);
        process.stdout.write(`Traitement du lot ${Math.floor(i / BATCH_SIZE) + 1} (Requêtes ${i + 1} à ${Math.min(i + BATCH_SIZE, pendingQueries.length)})...`);
        
        const results = await processBatch(ai, batch);
        
        if (results) {
            let batchModifs = 0;
            for (const query of catalog) {
                if (results[query.id]) {
                    query.description = results[query.id];
                    batchModifs++;
                    modificationsCount++;
                }
            }
            // SAUVEGARDE PROGRESSIVE
            fs.writeFileSync(catalogPath, JSON.stringify(catalog, null, 2) + '\n', 'utf-8');
            console.log(` Fait ! (${batchModifs} descriptions ajoutées)`);
        } else {
            console.log(" Échec de ce lot, on continue avec le suivant...");
        }
        
        // Petite pause de politesse
        await delay(5000);
    }

    console.log(`\nTerminé ! ${modificationsCount} nouvelles descriptions générées.`);
}

main().catch(console.error);

File: scripts/generate_ai_descriptions.py
import json
import os
import sys
import time
from pathlib import Path

try:
    from google import genai
except ImportError:
    print("Veuillez installer le package google-genai : pip install google-genai")
    sys.exit(1)

def generate_description(client: genai.Client, query: dict) -> str:
    prompt = f"""
Tu es un expert métier et technique sur l'ERP Sage X3.
Voici une requête SQL extraite de l'ERP. 
Titre de la requête : {query.get('title')}
Domaine métier : {query.get('domain')}
Tables utilisées : {", ".join(query.get('fields', []))}
Code SQL : 
{query.get('sql')}

Génère une description très courte (1 à 2 phrases maximum) en français qui explique simplement à quoi sert cette extraction de données. L'explication doit être compréhensible par un utilisateur métier.
Ne donne QUE la description finale, aucune introduction ni conclusion.
"""
    try:
        response = client.models.generate_content(
            model='gemini-2.5-flash',
            contents=prompt,
        )
        return response.text.strip()
    except Exception as e:
        print(f"\nErreur pour la requête {query.get('id')}: {e}")
        return ""

def main():
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("Erreur : La variable d'environnement GEMINI_API_KEY n'est pas définie.")
        print("Pour la définir sous PowerShell :")
        print("$env:GEMINI_API_KEY=\"votre_cle_api\"")
        sys.exit(1)

    client = genai.Client(api_key=api_key)
    
    catalog_path = Path("src/lib/data/queriesCatalog.json")
    if not catalog_path.exists():
        print(f"Erreur: Le fichier {catalog_path} est introuvable. Lancez le script depuis la racine du projet.")
        sys.exit(1)

    print("Chargement du catalogue des requêtes...")
    with open(catalog_path, "r", encoding="utf-8") as f:
        catalog = json.load(f)

    total = len(catalog)
    print(f"{total} requêtes trouvées. Début de la génération...\n")

    modifications_count = 0

    for i, query in enumerate(catalog):
        print(f"[{i+1}/{total}] Traitement de '{query.get('title')}' ({query.get('id')})...", end="", flush=True)
        
        # On ne traite que les descriptions par défaut
        if query.get("description", "").startswith("Extraction de données basée sur les tables"):
            new_desc = generate_description(client, query)
            if new_desc:
                query["description"] = new_desc
                modifications_count += 1
                print(" Fait.")
            else:
                print(" Échoué.")
            
            # Petite pause pour éviter de saturer l'API (Rate limiting)
            time.sleep(2)
        else:
            print(" Déjà décrit, ignoré.")

    print(f"\nSauvegarde de {modifications_count} nouvelles descriptions...")
    with open(catalog_path, "w", encoding="utf-8") as f:
        json.dump(catalog, f, ensure_ascii=False, indent=2)
        # Ajouter le saut de ligne final conventionnel
        f.write("\n")

    print("Génération terminée avec succès !")

if __name__ == "__main__":
    main()
